src/main/scala/akka/stream/contrib/ZipInputStreamSource.scala

```diff
@@ -0,0 +1,206 @@
+   *
+   * @param in a function that builds a [[ZipInputStream]]
+   * @param chunkSize the size of the chunks
+   * @return [[ZipInputStreamSource]] instance
+   */
+  def create(
+    in:        Function0[ZipInputStream],
+    chunkSize: Int                       = DefaultChunkSize
+  ): Source[(ZipEntryData, ByteString), Future[Long]] =
+    Source.fromGraph(new ZipInputStreamSource(in, chunkSize))
+      .withAttributes(name("zipInputStreamSource") and IODispatcher)
+}
+
+/**
+ * A stage that works as a [[Source]] of data chunks extracted from
+ * zip files. In addition to regular files, the zip file might contain
+ * directories and other zip files. Every chunk is a tuple of [[ZipEntryData]]
+ * and [[ByteString]], where the former carries basic info about the file from
+ * which the bytes come and the latter carries those bytes. This stage
+ * materializes to the total amount of read bytes.
+ *
+ * @param in a function that builds a [[ZipInputStream]]
+ * @param chunkSize the size of the chunks
+ */
+final class ZipInputStreamSource private (
+  in:        () => ZipInputStream,
+  chunkSize: Int                  = DefaultChunkSize
+) extends GraphStageWithMaterializedValue[SourceShape[(ZipEntryData, ByteString)], Future[Long]] {
+
+  val matValue = Promise[Long]()
+
+  override val shape =
+    SourceShape(Outlet[(ZipEntryData, ByteString)]("zipInputStreamSource.out"))
+
+  override def createLogicAndMaterializedValue(inheritedAttributes: Attributes) = {
+
+    val InputBuffer(initialBuffer, maxBuffer) =
+      inheritedAttributes.getAttribute(classOf[InputBuffer], InputBuffer(16, 16))
+
+    val logic = new GraphStageLogic(shape) {
+      import shape._
+
+      private var is: ZipInputStream = _
+      private var readBytesTotal: Long = 0L
+      private var buffer: Vector[(ZipEntryData, ByteString)] = Vector.empty
+      private var eof: Boolean = false
+      private var currentEntry: Option[ZipEntry] = None
+      private var currentStreams: Seq[ZipInputStream] = Seq()
+
+      override def preStart(): Unit = {
+        super.preStart()
+        try {
+          is = in()
+          currentStreams = List(is)
+          fillBuffer(initialBuffer)
+        } catch {
+          case ex: Exception =>
+            matValue.failure(ex)
+            failStage(ex)
+        }
+      }
+
+      setHandler(out, new OutHandler {
+
+        override def onPull(): Unit = {
+          fillBuffer(maxBuffer)
+          buffer match {
+            case Seq() =>
+              finalize()
+            case head +: Seq() =>
+              push(out, head)
+              finalize()
+            case head +: tail =>
+              push(out, head)
+              buffer = tail
+          }
+          def finalize() = try {
+            is.close()
+          } finally {
+            matValue.success(readBytesTotal)
+            complete(out)
+          }
+        }
+
+        override def onDownstreamFinish(): Unit =
+          try {
+            is.close()
```

Comment:
`close()` is very important when using GZIPInputStream for avoiding leaking off-heap memory. Should we add additional `close()` in `postStop`?